### ISSUE TITLE
Sampling polygon data with points using v.what.vect is slow and wrong

### ISSUE THREAD

**Issue by tschmetzer** [qgis/QGIS#46916]
### What is the bug or the crash?

Sampling 3.000 points from a polygon layer's decimal attribute results in 10 min calculation time and a point layer containing only NULL values

![grafik](https://user-images.githubusercontent.com/67278094/150365061-537b3a3a-173a-4612-87a4-9fd74e326743.png)

### Steps to reproduce the issue

1. Download and insert point layer into QGIS: [points_baleares_copy.zip](https://github.com/qgis/QGIS/files/7907090/points_baleares_copy.zip)

2. Download and insert polygon layer into QGIS: https://wri-projects.s3.amazonaws.com/Aqueduct30/finalData/Y2019M07D12_Aqueduct30_V01.zip: path in zip file: Y2019M07D12_Aqueduct30_V01\baseline\annual\y2019m07d11_aqueduct30_annual_v01.gpkg
![grafik](https://user-images.githubusercontent.com/67278094/150364517-2a83af52-7ef9-4cc6-b139-e2de129663f0.png)
3. Check the (empty) field 'rri_score' extists in point layer points_baleares
![grafik](https://user-images.githubusercontent.com/67278094/150365659-0ef39b73-3450-41c8-b982-df4c4bb0cf79.png)
4. Call plugin `v.what.vector` with options shown in the screenshot below
![grafik](https://user-images.githubusercontent.com/67278094/150368282-023c9267-2b0f-449e-8267-efa9dced292f.png)

5. Observe result:
![grafik](https://user-images.githubusercontent.com/67278094/150370843-2b952c41-887b-4454-9ba7-ae275f759785.png)

### Versions

QGIS version | 3.22.3-Białowieża | QGIS code revision | 1628765ec7
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.5
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
Compiled against GEOS | 3.10.0-CAPI-1.16.0 | Running against GEOS | 3.10.2-CAPI-1.16.0
SQLite version | 3.35.2
PDAL version | 2.3.0
PostgreSQL client version | 13.0
SpatiaLite version | 5.0.1
QWT version | 6.1.3
QScintilla2 version | 2.11.5
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
GroupStats | 2.2.5
pointsamplingtool | 0.5.3
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.5
processing | 2.12.99
sagaprovider | 2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by tschmetzer** [comment]
The processing algorithm "Join attributes by location" takes 32 seconds yielding a correct result.

**Comment by gioman** [comment]
> The processing algorithm "Join attributes by location" takes 32 seconds yielding a correct result.

@tschmetzer GRASS tools in processing can have an awful overhead because of the necessity to import and export on the fly the data to/from a temporary GRASS mapset/locations. About the tool not yelding results: I'm even surprised this tool is here, is one of that tools where GRASS updates a layer rather than creating a new one. Not sure of this has been implemented supported in GRASS/Processing, but in general it is expected for a (3rd party) tool to generate a new output.

**Comment by tschmetzer** [comment]
> > The processing algorithm "Join attributes by location" takes 32 seconds yielding a correct result.
> 
> @tschmetzer GRASS tools in processing can have an awful overhead because of the necessity to import and export on the fly the data to/from a temporary GRASS mapset/locations. About the tool not yelding results: I'm even surprised this tool is here, is one of that tools where GRASS updates a layer rather than creating a new one. Not sure of this has been implemented supported in GRASS/Processing, but in general it is expected for a (3rd party) tool to generate a new output.

I am a bit confused. The tool generates a new output given in "Updated".

Whatever the reason may be - many users (like me) may lack this information. If it's there either it should work or some specific information needs to be communicated to the user why it doesn't work or how to make it work. Well but first we need to understand the root cause..

By the way:  “y2019m07d11_aqueduct30_annual_v01” seems to have some invalid geometry. No idea if this might have an impact...

**Comment by gioman** [comment]
> I am a bit confused. The tool generates a new output given in "Updated".

Yes, processing always generates new outputs (well, if we exclude the one that can run "in place"), but in this case is the underlying GRASS tool that does not (it is one of the few examples in GRASS by the way), but Processing expects it. As said I'm not not sure if this scenario was implemented specifically for this GRASS tool and similar ones (so in this case should return valid results) or if it was just added with a description file, but no one really paid attention to how it works.



> If it's there either it should work

sure, and thanks for the report/heads up.

**Comment by gioman** [comment]
@tschmetzer it works fine with my own data, not testing yours.

**Comment by gioman** [comment]
> 1. Download and insert point layer into QGIS: [points_baleares.zip](https://github.com/qgis/QGIS/files/7906409/points_baleares.zip)

@tschmetzer empty zip

**Comment by tschmetzer** [comment]
> > 1. Download and insert point layer into QGIS: [points_baleares.zip](https://github.com/qgis/QGIS/files/7906409/points_baleares.zip)
> 
> @tschmetzer empty zip

sorry, just fixed.

**Comment by gioman** [comment]
> sorry, just fixed. Plese use new zip file: https://github.com/qgis/QGIS/files/7907090/points_baleares_copy.zip

@tschmetzer a few things:

1) your layer have diiferent CRSs, for SAGA, GDAL, GRASS, etc they must have the same CRS

2) you polygon layer is enormous, it covers the whole world when your sampling points cover only a few polygons. As said GRASS tools have a overhead (because of the in/out) in GRASS data model. So you have two choices: or you run the tool using "only selected features" or you create a copy with only what you need. Also the layer has a really long list of columns, this also has in impact in the in/out operation.

Bottom line, it works.

[sampled.zip](https://github.com/qgis/QGIS/files/7907427/sampled.zip)

**Comment by tschmetzer** [comment]
Fixing the geometry didn't make any difference. I haven't tried your hints yet.

My understanding:
1. I guess the empty values (NULL) are due to different CRS. 
2. The enormous amount of data that is handled inefficiently between QGIS and the other interfaces is causing the long execution time.

I will try to convert the CRS tomorrow and check if I get it to work. No idea how CRS conversion works. I achieved to avoid that topic up to now :)

**Comment by nyalldawson** [comment]
I don't think there's any issue here (at least, not in QGIS). The grass tool is exposed and working as expected, and we have alternative high-performance tools available from the native provider which the user is already (subtely) steered to use.

**Comment by gioman** [comment]
> and we have alternative high-performance tools available from the native provider which the user is already (subtely) steered to use.

The GRASS tool are not slower per se, it is the overhead for in/out data to/from the temp mapset/location that we have in Processing that it can turns them slower if the datasets are very large. In fact arguably many GRASS tools are pretty fast when run against data that have been already imported in a GRASS location/mapset.

**Comment by tschmetzer** [comment]
Sorry for having captured the subtle steering ;)

I re-ran the data with a CRS converted layer and a smaller subset of the overlay layer

Committed to improve QGIS and the prevention of reoccurring mistakes by other users I suggest to include a hint for all concerned external tools such as SAGA, GDAL, GRASS, etc with a statement like:

```
Note:
External tools usually expect coherent CRSs and have a larger overhead due to temporary data transfer. 
It is recommended to convert to equal input CRSs and use the smallest data subset possible. 
Otherwise the processing might lead to supposedly wrong results and long processing times.
```

The text is subject to discussion of course and should be as concise as possible.

Example illustration: 
![image](https://user-images.githubusercontent.com/67278094/150529686-0b27df83-5604-4d0c-87ba-2afc4b7abb70.png)

Shall we reopen the ticket as a feature request?

**Comment by gioman** [comment]
@tschmetzer there is already an option in Processing configs

![image](https://user-images.githubusercontent.com/1951107/150530808-892bcd28-ca80-4f8d-8959-66cfe7575391.png)


if it does not work then it should warrant a ticket.

**Comment by tschmetzer** [comment]
Is this not default? Cannot remember to have it turned off. I will check later.

**Comment by tschmetzer** [comment]
> Is this not default? Cannot remember to have it turned off. I will check later.

I checked: The option was and is turned on. Yet I have not been warned, even when I just now tested again.
What should the warning have looked like? Why did it it not show up?

**Comment by gioman** [comment]
> What should the warning have looked like? Why did it it not show up?

it used to work (blocking the operation or raising a warning, can't remember clearly). Feel free to file a ticket (please first search of it already exists).
